test/test_nn.py

```diff
@@ -352,7 +352,7 @@ def _backward_criterion(self, criterion, input, target, gradOutput=None, extra_a
         args = input_tuple + (target,) + extra_args
         if gradOutput is None:
             gradOutput = torch.ones(())
-        criterion(*args).backward(gradOutput.to(input_tuple[0]))
+        criterion(*args).backward(gradOutput.to(output))
```

Comment:
For C to R functions, input's dtype is not equal to output's dtype. We need gradoutput to be the same dtype as output.

Reply:
hmm what if the output is a tuple? I think we should add a similar check for output as input:
`output_tuple = output if isinstance(output, tuple) else (output,)`

Reply:
Can output be a tuple though? Input might be a tuple only because when we backward, we might want to populate the grads of multiple inputs. I'm curious which functions return tuples.

Reply:
not sure about `torch.nn` module functions, but some torch functions that come to mind are `triangular_solve`, `qr`

Reply:
@soulitzer did you look into this?

Reply:
There are functions like `nn.AdaptiveMaxPool2d` that do return a tuple, so I ended up adding the check for the tuple case.